test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul

```diff
@@ -477,15 +477,16 @@
 //             pos := add(pos, 0x60)
 //         }
 //         let _3 := mload(64)
-//         if slt(sub(_3, length), 128) { revert(_1, _1) }
-//         let offset := calldataload(add(length, 64))
-//         let _4 := 0xffffffffffffffff
-//         if gt(offset, _4) { revert(_1, _1) }
-//         let value2 := abi_decode_t_array$_t_uint256_$dyn_memory_ptr(add(length, offset), _3)
-//         let offset_1 := calldataload(add(length, 0x60))
-//         if gt(offset_1, _4) { revert(_1, _1) }
-//         let value3 := abi_decode_t_array$_t_array$_t_uint256_$2_memory_$dyn_memory_ptr(add(length, offset_1), _3)
-//         sstore(calldataload(length), calldataload(add(length, 0x20)))
+//         let _4 := mload(0x20)
+//         if slt(sub(_3, _4), 128) { revert(_1, _1) }
```

Comment:
It seems suspicious that `slt(sub(_3, length), 128)` was replaced by `slt(sub(_3, _4), 128)` because

- `length` contains the value at memory location whose value is at memory location 0x0 (i.e., `length` is obtained by dereferencing the "pointer" stored at memory location 0x0)
- but `_4` contains the value at memory location 0x20

It is suspicious because what is contained in the location whose value is at 0x0 can be completely different from the value at memory location 0x20, unless I am missing something.

Reply:
there might have been multiple layers of mload in between. This code is the result of heavy inlining.

Reply:
Ah, I think I now understand. On this line

https://github.com/ethereum/solidity/blob/ae90aaceab81650de9cc0a2073c4048666b2bf15/test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul#L469

`length` gets written to memory location 0x20. After `length` gets written to this memory location, there is a function call that contains a for loop that contains an mstore. See

https://github.com/ethereum/solidity/blob/ae90aaceab81650de9cc0a2073c4048666b2bf15/test/libyul/yulOptimizerTests/fullSuite/abi_example1.yul#L550

Therefore, we can no longer assume that what was written to 0x20 earlier (`length`) is still intact. If I follow the code, then whatever is at 0x20 is the "new length". I think this is the reason, we've had to swap `length` by `mload(0x20)` i.e., `_4`

Reply:
Sounds good!